core/src/main/java/org/apache/calcite/rel/rules/MinusToFilterRule.java

```diff
@@ -0,0 +1,137 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to you under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.calcite.rel.rules;
+
+import org.apache.calcite.plan.RelOptRuleCall;
+import org.apache.calcite.plan.RelOptUtil;
+import org.apache.calcite.plan.RelRule;
+import org.apache.calcite.rel.RelNode;
+import org.apache.calcite.rel.core.Filter;
+import org.apache.calcite.rel.core.Minus;
+import org.apache.calcite.rex.RexNode;
+import org.apache.calcite.rex.RexUtil;
+import org.apache.calcite.tools.RelBuilder;
+
+import org.immutables.value.Value;
+
+/**
+ * Rule that replaces {@link org.apache.calcite.rel.core.Minus}
+ * operator with {@link org.apache.calcite.rel.core.Filter}
+ * when both inputs are from the same source with only filter conditions differing.
+ *
+ * <p>Example transformation:
+ * <blockquote><pre>
+ * SELECT mgr, comm FROM emp WHERE mgr = 12
+ * EXCEPT
+ * SELECT mgr, comm FROM emp WHERE comm = 5
+ *
+ * to
+ *
+ * SELECT DISTINCT mgr, comm FROM emp
+ * WHERE mgr = 12 AND NOT(comm = 5)
+ * </pre></blockquote>
+ */
+@Value.Enclosing
+public class MinusToFilterRule
+    extends RelRule<MinusToFilterRule.Config>
+    implements TransformationRule {
+
+  /** Creates an MinusToFilterRule. */
+  protected MinusToFilterRule(Config config) {
+    super(config);
+  }
+
+  //~ Methods ----------------------------------------------------------------
+
+  @Override public void onMatch(RelOptRuleCall call) {
+    final Minus minus = call.rel(0);
+    if (minus.all || minus.getInputs().size() != 2) {
+      return;
+    }
+
+    final RelNode leftInput = minus.getInput(0);
+    final RelNode rightInput = minus.getInput(1);
+
+    final RelBuilder builder = call.builder();
+
+    final RelNode leftBase = getBaseSource(leftInput);
+    final RelNode rightBase = getBaseSource(rightInput);
+
+    RexNode rightCond = extractFilterCond(rightInput);
+    if (!RexUtil.isDeterministic(rightCond)) {
+      return;
+    }
+
+    // Verify they come from the same source
+    if (!leftBase.equals(rightBase)) {
+      return;
+    }
+
+    final RexNode finalCond =
+        builder.and(extractFilterCond(leftInput), builder.not(rightCond));
```

Comment:
where is the check that both filters produce the same fields?

Reply:
Yes, by identifying the beginning and ending double quotes.

Reply:
I understand that there is no need to judge the same column, as long as the condition of `right` is reversed. WDYT

Reply:
How about `SELECT mgr, comm FROM emp WHERE mgr = 12 EXCEPT SELECT mgr1, comm1 FROM emp WHERE comm = 5`? We need to ensure that the two inputs return the same columns. Only then will this Rule be valid.

Reply:
Maybe [here](https://github.com/apache/calcite/pull/4322/files/1a743cc2f3947a79a084bd1fe433a2740961a2cf#diff-7dde415fabdd836a8181e96f6405a572bd5a9c07d67394d596dbbd8039e1454aR80) can solve this question？I will check if it comes from the same source.

Reply:
Get it.